### PR TITLE
Updating OkHttp to the latest version

### LINKED ISSUE #195: Java SDK:  Set up to handle ping's as required/desired
We've now seen a few cases where connectors that are not terribly active get timed out by the server.  Internally, they should handle reconnect, but it might be good to avoid doing this.

OkHttp3 (which we use for our WS implementation) can create a client that will automatically perform ping processing (`OkHttpClient.Builder()...pingInterval(duration)`).  This might require that we upgrade the OkHttp3 version we use (we currently use 3.4, that's up to 3.14 (I think), and 4.x).

We may want to do this as an option -- "active" connectors don't need it and it'll just increase communication.  But if there's a type of connector that gets work to do (or work that requires communication) rarely, then it may be a good thing for which to provide.

### LINKED ISSUE #204: Add keepAlive ping to connector SDK
Depending on the cloud provider, an idle connection between the connector and VANTIQ may get closed on some time interval, forcing a reconnection. In some cases, we've seen a bug on the VANTIQ side claiming a "connection already exists" for the given source. This issue seems to occur fairly randomly, and I have not been able to find a consistently reproducible case.

Adding some keepAlive functionality in the underlying connector SDK could prevent that scenario from ever arising, because the connection would never go idle. The VANTIQ server has the ability to respond to "ping" messages, so we'd only need to update the SDK code.

### LINKED ISSUE #260: Update our version of OkHttp
Update the OkHttp library to at least verson 3.10.0 so that we can pickup the `pingInterval` method. We can then offer SDK overrides to utilize the `pingInterval` if desired.

### PR DESCRIPTION
**Pull request by namirfa**
Closes #260 
Closes #204 
Closes #195

Updates the OkHttp library to the latest stable release. Since version 3.5, OkHttp has folded in the separate `-ws` package into the general okhttp one. The majority of these changes convert our old WebSocket implementations to use the newer version.

This PR also adds the ability to configure the OkHttp client to send pings to the Vantiq Server, which can be specified by including a `sendPings` property in the `server.config` file.

### CODE REVIEW

**Comment by namirfa** [comment]
[Jenkins Build](https://jenkins.vantiq.com/jenkins/job/vantiq-extension-sources-branch/21/)

**Comment by fhcarter** [comment]
Shouldn't we add the `sendPing` documentation here?

```
targetServer=https://dev.vantiq.com/
```
								
[extjsdk/README.md](https://github.com/vantiq/vantiq-extension-sources/blob/4a1a713168cb2a6b8226d4edd03024556a101808/extjsdk/README.md#L61) (Line 61)

**Comment by namirfa** [comment]
> Shouldn't we add the `sendPing` documentation here?
> 
> ```
> targetServer=https://dev.vantiq.com/
> ```
> 
> [extjsdk/README.md](https://github.com/vantiq/vantiq-extension-sources/blob/4a1a713168cb2a6b8226d4edd03024556a101808/extjsdk/README.md#L61) (Line 61)

Yep I'm adding that as we speak!

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java [hunk 1]
```diff
@@ -511,9 +505,21 @@ public boolean receivedMessage() {
         }
 
         @Override
-        public void sendPing(Buffer payload) throws IOException {}
+        public boolean close(int code, String reason) {
+            client.getListener().onClosed(null, code,reason);
```
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java:509]: Seems like instead of `null` for the WebSocket, we can pass in `this` can't we?  Seems like that would be cleaner, since it is a `WebSocket`.

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java [hunk 2]
```diff
@@ -511,9 +505,21 @@ public boolean receivedMessage() {
...
+            return true;
+        }
+
+        //================================ Necessary to implement WebSocket ================================
+
+        @Override
+        public void cancel() {}
+        @Override
+        public long queueSize() {return 0;}
```
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java:518]: Same stuff about the multiline things to keep Sider happy.

**Comment by namirfa** [comment]
@fhcarter
Documented. If it's alright with you, I'm going to create a separate issue for the TODO I added in the `Utils` class.

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java [hunk 3]
```diff
@@ -511,9 +505,21 @@ public boolean receivedMessage() {
...
+        @Override
+        public Request request() {return null;}
```
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java:520]: ditto the sider stuff & multiline

File: extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketListener.java [hunk 4]
```diff
@@ -212,25 +209,13 @@ public void onOpen(WebSocket webSocket, okhttp3.Response response) {
      * Translate the received message and pass it on to the related handler. Additionally, updates the client about
      * successful authentications and source connections.
      *
-     * @param body  The {@link ResponseBody} containing the message received.
+     * @param webSocket The websocket on which the message was received.
+     * @param bodyBytes  The {@link ByteString} containing the message received.
      */
     @Override
-    public void onMessage(ResponseBody body) {
+    public void onMessage(WebSocket webSocket, ByteString bodyBytes) {
```
Comment (fhcarter) [inline extjsdk/src/main/java/io/vantiq/extjsdk/ExtensionWebSocketListener.java:216]: In the IDE, all the `on*()` calls seem to have a complaint from the IDE that your `@Override`s aren't annotated the same way as their super-methods.  Most are missing the `@NonNull`.

Probably better to keep these clean though it may cause issues with the tests that try to pass in `null` `WebSocket` instances a lot.  Probably better to just pass in the appropriate `FalseWebSocket`.

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketListener.java [hunk 5]
```diff
@@ -115,7 +116,7 @@ public void testSourceConnection() {
     public void testSourceConnectionFailFirst() {
         authenticate(true);
         
-        listener.onMessage(TestListener.errorMessage());
+        listener.onMessage(null, TestListener.errorMessage());
```
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketListener.java:119]: Sending the actual web socket here would be much preferable...

File: extjsdk/src/main/java/io/vantiq/extjsdk/Utils.java [hunk 6]
```diff
@@ -84,4 +85,30 @@ public static Integer obtainTCPProbePort() {
...
+            } else {
+                return false;
+            }
+        } catch (IOException e) {
+            log.error("An error occurred while trying to retrieve the sendPings value.", e);
+            return false;
+        }
+    }
+
+    // TODO NAMIR - Add helper that returns the properties object since that code is repeated
```
[the remark this answers is not in the document]
Comment (fhcarter) [reply extjsdk/src/main/java/io/vantiq/extjsdk/Utils.java:113]: Can be separate issue, I suppose. But since we support naming your config file, that support should extend here, too.

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java (hunk 2 above)
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketClient.java:518]: ditto the sider stuff & multiline things

File: extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketListener.java (hunk 5 above)
Comment (fhcarter) [inline extjsdk/src/test/java/io/vantiq/extjsdk/TestExtensionWebSocketListener.java:119]: May have to extend the client or listener to return the `FalseWebSocket` they use to pass in the right thing here.

File: extjsdk/README.md [hunk 7]
```diff
@@ -47,6 +46,10 @@ The SDK includes a utility method to retrieve the startup configuration document
 `<connectorWorkingDirector>/serverConfig/server.config`. The SDK will also look for the file in the working connector 
 directory if it is not found in the `serverConfig` subdirectory (i.e. `<connectorWorkingDirector>/server.config`).
 
+In addition to the minimum configuration properties, the `server.config` file can also include the following property:
+*   `sendPings`: A boolean property that, if set to `true`, enables the SDK to send ping messages to the Vantiq Server.
```
Comment (fhcarter) [inline extjsdk/README.md:50]: To keep all the MD processors happy, usually a good idea to surround bullet lines with blank lines.

**Review by fhcarter — CHANGES_REQUESTED** [review]
Mostly good.  I think there are a few places where it makes sense to clean up the sider warnings (keeps with our internal coding style anyway).  Also, I tend to think it'd be better in the tests to pass the websocket thru (rather than null).  Doesn't do that much, but allows the various `on` handlers to be properly declared (which the IDE is whining about anyway).

**Comment by namirfa** [comment]
@fhcarter Added fixes per your comments.

[Latest Jenkins Build](https://jenkins.vantiq.com/jenkins/job/vantiq-extension-sources-branch/22/)

Manually ran tests with server, which are all passing.

**Comment by namirfa** [comment]
I did not add the `Utils` fix yet. Those methods should definitely accept a filename parameter, but I agree with your comment that it may be better to refactor those Utils in general to only read the `server.config` once, (and offer a `clear()` method as well). I'd like to make those changes in a separate PR if that's alright.

**Comment by fhcarter** [comment]
I think that's OK, but we should try to do it soon.  It leaves the new stuff in a sort-of broken state (must used default file name) technically.  I don't think any of that is used yet, but still...

**Review by fhcarter — APPROVED** [review]
LGTM functionally.

There's one issue we should fix (`@NonNull` parameters) and a bunch of code style (missing/extra blank lines) issues.  But you can fix those & merge w/o another review.

Good to get the sider complaints done.  I'm sort of on a push to try & deal with those.  I wasn't impressed before, but some things are useful. (others not, but I'm still working on how to most easily disable the things we don't care about so we don't have to annotate all the files).
